feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,206 @@
+	config        *n7.N7ClientConfig
+	notifyServer  *echo.Echo
+	cloudRegistry service_registry.GatewayRegistry
+}
+
+func NewNotificationHandler(
+	config *n7.N7ClientConfig,
+	cloudReg service_registry.GatewayRegistry,
+) *NotificationHandler {
+	return &NotificationHandler{
+		config:        config,
+		notifyServer:  echo.New(),
+		cloudRegistry: cloudReg,
+	}
+}
+
+// RegisterHandlers registers the UpdateNotification and TerminateNotification handlers.
+// The notification url is of the form {notifyRoot}/{encodedSessionId}/{operation}
+//   Example: http://magma-feg.magma.com/sm-policy-control/v1/MTIzNDU2Nzg5MDsxMjM0NQo=/update
+//   Here notifyRoot = http://magma-feg.magma.com/sm-policy-control/v1
+//        encodedSessionId = MTIzNDU2Nzg5MDsxMjM0NQo= (Session-Id is urlencoded)
+//        operation = update
+// This notification url is send to PCF in the SmPolicyCreate request
+func (handler *NotificationHandler) RegisterHandlers() error {
+	urlDef, err := url.ParseRequestURI(handler.config.NotifyApiRoot)
+	if err != nil {
+		return fmt.Errorf("error parsing notify api root - %s", err)
+	}
+	basePath := fmt.Sprintf("%s/:%s", urlDef.Path, EncodedSessionId)
+	// Register the handlers
+	handler.notifyServer.POST(basePath+"/update", handler.postSmPolicyUpdateNotification)
+	handler.notifyServer.POST(basePath+"/terminate", handler.postSmPolicyTerminateNotification)
+	return nil
+}
+
+func (handler *NotificationHandler) Start() (string, error) {
+	errChan := make(chan error)
+	addr, err := net.ResolveTCPAddr("tcp", handler.config.LocalAddr)
+	if err != nil {
+		return "", err
+	}
+
+	go func() {
+		err = handler.notifyServer.Start(addr.String())
+		if err != nil {
+			errChan <- err
+		}
+	}()
+
+	return sbi.WaitForSbiServerAndGetAddr(handler.notifyServer, errChan)
+}
+
+func (handler *NotificationHandler) Shutdown() {
+	ctx, cancel := context.WithTimeout(context.Background(), ShutdownTimeout)
+	defer cancel()
+	handler.notifyServer.Shutdown(ctx)
+}
+
+// postSmPolicyUpdateNotification handles SM policy update notification request from PCF.
+func (handler *NotificationHandler) postSmPolicyUpdateNotification(ctx echo.Context) error {
+	client, err := relay.GetSessionProxyResponderClient(handler.cloudRegistry)
+	if err != nil {
+		glog.Errorf("postSmPolicyUpdateNotification failed to get SessionProxyResponderClient: %s", err)
+		return fmt.Errorf("internal server error")
+	}
+	defer client.Close()
+
+	var smUpdateNotify n7_client.SmPolicyNotification
+	err = ctx.Bind(&smUpdateNotify)
+	if err != nil {
+		err = fmt.Errorf("invalid SmPolicyNotification received: %s", err)
+		glog.Errorf("postSmPolicyUpdateNotification: %s", err)
+		return echo.NewHTTPError(http.StatusBadRequest, err)
+	}
+	sessionId, imsi, err := getSessionIdAndIMSI(ctx.Param(EncodedSessionId))
+	if err != nil {
+		glog.Errorf("postSmPolicyUpdateNotification unable to fetch session-id for UpdateNotify - %s", err)
+		return echo.NewHTTPError(http.StatusBadRequest, err)
```

Comment:
why some returns are `echo.NewHTTPErro` and some others are not?

Check all of them